File: practice.py
n = int(input("Enter no. of processes: "))

pid = []
at = []
bt = []

for i in range(n):
    pid.append(i + 1)
    at.append(int(input(f"Enter arrival time for P{i+1}: ")))
    bt.append(int(input(f"Enter burst time for P{i+1}: ")))

# Sort processes by arrival time: (pid, at, bt)
processes = sorted(zip(pid, at, bt), key=lambda x: x[1])

ct = []   # Completion Time
tat = []  # Turnaround Time
wt = []   # Waiting Time
rt = []   # Response Time
pr = []   # Penalty/Performance Ratio = TAT / BT

time = 0

for p in processes:
    pid_i, at_i, bt_i = p

    # Start time is max(current time, arrival time)
    start_time = max(time, at_i)

    # Response time = first CPU start - arrival
    rt_i = start_time - at_i

    # Completion time = start time + burst
    time = start_time + bt_i
    ct_i = time

    tat_i = ct_i - at_i            # Turnaround time
    wt_i = tat_i - bt_i            # Waiting time
    pr_i = tat_i / bt_i            # Penalty/Performance ratio

    ct.append(ct_i)
    tat.append(tat_i)
    wt.append(wt_i)
    rt.append(rt_i)
    pr.append(pr_i)

print("\nPID\tAT\tBT\tCT\tTAT\tWT\tRT\tPR")

for i in range(n):
    print(f"{processes[i][0]}\t"
          f"{processes[i][1]}\t"
          f"{processes[i][2]}\t"
          f"{ct[i]}\t"
          f"{tat[i]}\t"
          f"{wt[i]}\t"
          f"{rt[i]}\t"
          f"{pr[i]:.2f}")

# Averages
avg_ct = sum(ct) / n
avg_tat = sum(tat) / n
avg_wt = sum(wt) / n
avg_rt = sum(rt) / n
avg_pr = sum(pr) / n

# Throughput = number of processes / total time (from first arrival to last completion)
first_arrival = processes[0][1]
last_completion = ct[-1]
total_time = last_completion - first_arrival
throughput = n / total_time if total_time > 0 else 0

print("\nAverage CT  =", avg_ct)
print("Average TAT =", avg_tat)
print("Average WT  =", avg_wt)
print("Average RT  =", avg_rt)
print("Average PR  =", avg_pr)
print("Throughput  =", throughput, "processes/unit time")

File: practice.java
import java.util.Scanner;

public class practice {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter no. of processes: ");
        int n = sc.nextInt();

        int pid[] = new int[n];
        int at[] = new int[n];
        int bt[] = new int[n];
        int ct[] = new int[n];
        int tat[] = new int[n];
        int wt[] = new int[n];

        // inputs
        for(int i = 0; i < n; i++){
            System.out.print("AT for P" + (i + 1) + ": ");
            pid[i] = i+1;
            at[i] = sc.nextInt();
        }

        for(int i = 0; i < n; i++){
            System.out.print("BT for P" + (i + 1) + ": ");
            bt[i] = sc.nextInt();
        }

        // sorting
        for(int i = 0; i < n - 1; i++){
            for(int j = 0; j < n - i - 1; j++){
                if(at[j] > at[j + 1]){
                    int temp = at[j];
                    at[j] = at[j+1];
                    at[j+1] = temp;

                    temp = bt[j];
                    bt[j] = bt[j+1];
                    bt[j+1] = temp;

                    temp = pid[j];
                    pid[j] = pid[j+1];
                    pid[j+1] = temp;
                }
            }
        }

        // calculate CT, TAT, WT

        int time = 0;
        float sumTAT = 0, sumWT = 0;

        for(int i = 0; i < n; i++){
            if(at[i] > time){
                time = at[i];
            }

            time += bt[i];
            ct[i] = time;

            tat[i] = ct[i] - at[i];
            wt[i] = tat[i] - bt[i];

            sumTAT += tat[i];
            sumWT += wt[i];
        }

        // output table

        System.out.println();
        System.out.println("PID\tAT\tBT\tCT\tTAT\tWT");
        for(int i = 0; i < n; i++){
            System.out.printf("P%-2d\t%-2d\t%-2d\t%-2d\t%-2d\t%-2d%n", pid[i], at[i], bt[i], ct[i], tat[i], wt[i]);
        }

        System.out.println();

        System.out.println("Avg. TAT: " + sumTAT/n);
        System.out.println("Avg. WT: " + sumWT/n);



        sc.close();
    }
}